File: src/main/scala/realTimeProcess.scala
import org.apache.spark.sql.SparkSession
from pyspark.sql import SparkSession
import org.apache.spark.SparkConf

object realTimeProcess {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
      .setAppName("Simple Application")
      .setMaster("local")

    val spark = SparkSession.builder.appName("test").config("spark.sql.debug.maxToStringFields", "100")
      .config(conf)
      .config("hive.metastore.uris", "thrift://localhost:9083")
      .config("hive.metastore.client.factory.class", "com.amazonaws.glue.catalog.metastore.AWSGlueDataCatalogHiveClientFactory")
      .getOrCreate()

    val dsraw = spark
      .readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", "172.42.71.50:29092")
      .option("subscribe", "dbserver1.inventory.orders")
      .option("startingOffsets", "earliest")
      .load()


   val ds = dsraw.selectExpr("CAST(value AS STRING)")


  }
}





# kafka
dsraw = spark \
.readStream \
.format("kafka") \
.option("kafka.bootstrap.servers", "172.42.71.50:29092") \
.option("subscribe", "dbserver1.inventory.orders") \
.option("startingOffsets", "earliest") \
.load()


ds = dsraw.selectExpr("CAST(value AS STRING)")

File: src/main/python/realTimeProcess.py
import logging

from pyspark.conf import SparkConf
from pyspark.sql import SparkSession

# def func_call(df):
#     df.selectExpr("CAST(value AS STRING) as json")
#     requests = df.rdd.map(lambda x: x.value).collect()
#     logging.info(requests)

spark = SparkSession \
    .builder \
    .appName("hasura") \
    .config("spark.sql.debug.maxToStringFields", "100") \
    .config("hive.metastore.uris", "thrift://localhost:9083", conf=SparkConf()) \
    .config("hive.metastore.client.factory.class",
            "com.amazonaws.glue.catalog.metastore.AWSGlueDataCatalogHiveClientFactory") \
    .getOrCreate()

#STEP1 subscribe to kafka topic
subscribe_df = spark \
    .readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", "localhost:29092") \
    .option("subscribe", "hasura.training.dataset") \
    .option("startingOffsets", "earliest") \
    .load()

#STEP2 create dataframe from log data

ds = subscribe_df.selectExpr("CAST(value AS STRING)")

#STEP3 clean data and apply transformation

apply_transformation_df = ds \
    .writeStream \
    .queryName("apply_transformation_query") \
    .format("memory") \
    .start()

result_df = spark.sql("select * from apply_transformation_query")

result_df.show()

#STEP4 write df to target location
result_df.write \
    .format('parquet') \
    .mode('append') \
    .save('s3://datadump/unsaved/streaming/')

